MON-3172: describe monitoring services (#2188)
openshift/cluster-monitoring-operator: Documentation/resources.md

## Routes

### openshift-monitoring/alertmanager-main

Expose the `/api` endpoints of the `alertmanager-main` service via a router.

### openshift-monitoring/prometheus-k8s

Expose the `/api` endpoints of the `prometheus-k8s` service via a router.

### openshift-monitoring/prometheus-k8s-federate

Expose the `/federate` endpoint of the `prometheus-k8s` service via a router.

### openshift-user-workload-monitoring/federate

Expose the `/federate` endpoint of the `prometheus-user-workload` service via a router.

### openshift-monitoring/thanos-querier

Expose the `/api` endpoints of the `thanos-querier` service via a router.

### openshift-user-workload-monitoring/thanos-ruler

Expose the `/api` endpoints of the `thanos-ruler` service via a router.

## Services

### openshift-monitoring/prometheus-operator-admission-webhook

Expose the admission webhook service which validates PrometheusRules and AlertmanagerConfig custom resources on port 8443.

### openshift-user-workload-monitoring/alertmanager-user-workload

Expose the user-defined Alertmanager web server within the cluster on the following ports:

> - Port 9095 provides access to the Alertmanager endpoints. Access requires a `RoleBinding` to the `monitoring-alertmanager-api-reader` role (for read-only operations) or `monitoring-alertmanager-api-writer` for the `openshift-user-workload-monitoring` project.

Author: a RoleBinding associates a Role (set of permissions) to a user (or service account).

> - Port 9095 provides access to the Alertmanager endpoints. Access requires a `RoleBinding` to the `monitoring-alertmanager-api-reader` role (for read-only operations) or `monitoring-alertmanager-api-writer` for the `openshift-user-workload-monitoring` project.

Author: For reference, the OCP docs say something like
> Assign the xxx role to a user in the openshift-user-workload-monitoring project.

> - Port 9095 provides access to the Alertmanager endpoints. Access requires a `RoleBinding` to the `monitoring-alertmanager-api-reader` role (for read-only operations) or `monitoring-alertmanager-api-writer` for the `openshift-user-workload-monitoring` project.

Author: idea?
```suggestion
- Port 9095 provides access to the Alertmanager endpoints. Access requires granting the `monitoring-alertmanager-api-reader` role (for read-only operations) or `monitoring-alertmanager-api-writer` role for the `openshift-user-workload-monitoring` project.
```

### openshift-monitoring/kube-state-metrics

Expose kube-state-metrics `/metrics` endpoints within the cluster on the following ports:
- Port 8443 provides access to the Kubernetes resource metrics. This port is for internal use, and no other usage is guaranteed.
- Port 9443 provides access to the internal kube-state-metrics metrics. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/metrics-server

Expose the metrics-server web server on port 443. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/monitoring-plugin

Expose the monitoring plugin service on port 9443. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/node-exporter

Expose the `/metrics` endpoint on port 9100. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/openshift-state-metrics

Expose openshift-state-metrics `/metrics` endpoints within the cluster on the following ports:
- Port 8443 provides access to the OpenShift resource metrics. This port is for internal use, and no other usage is guaranteed.

> - Port 9443 provides access to the internal openshift-state-metrics metrics. This port is for internal use, and no other usage is guaranteed.

Reviewer: ```suggestion
- Port 9443 provides access to the internal `openshift-state-metrics` metrics. This port is for internal use, and no other usage is guaranteed.
```

### openshift-monitoring/prometheus-adapter

> Expose the prometheus-adapter web server on port 443. This port is for internal use, and no other usage is guaranteed.

Reviewer: ```suggestion
Expose the `prometheus-adapter` web server on port 443. This port is for internal use, and no other usage is guaranteed.
```

### openshift-monitoring/prometheus-k8s

Expose the Prometheus web server within the cluster on the following ports:

> - Port 9091 provides access to all the Prometheus endpoints. Access requires a `ClusterRoleBinding` to the `cluster-monitoring-view` cluster role.

Reviewer: ```suggestion
- Port 9091 provides access to all the Prometheus endpoints. Access requires adding a `ClusterRoleBinding` [permission] to the `cluster-monitoring-view` cluster role.
```

- Port 9092 provides access to the `/metrics` endpoint only. This port is for internal use, and no other usage is guaranteed.

> 

Reviewer: ```suggestion
```

### openshift-user-workload-monitoring/prometheus-operator

Expose the `/metrics` endpoint on port 8443. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/prometheus-operator

Expose the `/metrics` endpoint on port 8443. This port is for internal use, and no other usage is guaranteed.

### openshift-user-workload-monitoring/prometheus-user-workload

Expose the Prometheus web server within the cluster on the following ports:
- Port 9091 provides access to the `/metrics` endpoint only. This port is for internal use, and no other usage is guaranteed.

> - Port 9092 provides access to the `/federate` endpoint only. Access requires a `ClusterRoleBinding` to the `cluster-monitoring-view` cluster role.

Reviewer: ```suggestion
- Port 9092 provides access to the `/federate` endpoint only. Access requires adding a `ClusterRoleBinding` [permission] to the `cluster-monitoring-view` cluster role.
```

> Also expose the `/metrics` endpoint of the Thanos sidecar web server on port 10902. This port is for internal use, and no other usage is guaranteed.

Reviewer: ```suggestion
This also exposes the `/metrics` endpoint of the Thanos sidecar web server on port 10902. This port is for internal use, and no other usage is guaranteed.
```

### openshift-monitoring/telemeter-client

Expose the `/metrics` endpoint on port 8443. This port is for internal use, and no other usage is guaranteed.

### openshift-monitoring/thanos-querier

Expose the Thanos Query web server within the cluster on the following ports:

> - Port 9091 provides access to all the Thanos Query endpoints. Access requires a `ClusterRoleBinding` to the `cluster-monitoring-view` cluster role.

Reviewer: ```suggestion
- Port 9091 provides access to all Thanos Querier endpoints. Access requires adding a `ClusterRoleBinding` [permission?] to the `cluster-monitoring-view` cluster role.
```

> - Port 9092 provides access to the `/api/v1/query`, `/api/v1/query_range/, `/api/v1/labels`, `/api/v1/label/*/values`, and `/api/v1/series` endpoints restricted to a given project. Access requires a `RoleBinding` to the `view` role for the project.

Reviewer: ```suggestion
- Port 9092 provides access to the `/api/v1/query`, `/api/v1/query_range/, `/api/v1/labels`, `/api/v1/label/*/values`, and `/api/v1/series` endpoints restricted to a given project. Access requires adding a `RoleBinding` [permission?] to the `view` role for the project.
```

> ## Routes

Author: @bburt-rh this is a very early draft showing how the "openshift.io/description" annotations can be translated into a MarkDown page. Feel free to edit/comment this file and I'll incorporate the changes to the source code.

Like we do for the Documentation/api.md, generating an asciidoc version shouldn't be complicated.

> ## Routes

Author: I imagine that we'll need some introduction explaining the goal of this page.

```
This document explains the following resources deployed and managed by the Cluster Monitoring Operator:
* ClusterRoles
* Roles
* Routes
* Services
```

Reviewer: Also, per the request in the docs Jira, I think we should add a warning about accessing API endpoints with similar info to what we include for the `/federate` endpoint: https://docs.openshift.com/container-platform/4.14/monitoring/accessing-third-party-monitoring-apis.html#monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus_accessing-third-party-monitoring-apis.

I'd think it would be especially important to warn about overloading the endpoint with queries.

WDYT?

Reviewer: Let's add the following, and make the 4 items in the list hyperlinks to the separate sections:

*******

This document describes the following resources deployed and managed by the Cluster Monitoring Operator (CMO):

* ClusterRoles
* Roles
* Routes
* Services

Important!

In certain situations, accessing endpoints can degrade the performance and scalability of your cluster, especially if you use endpoints to retrieve, send, or query large amounts of metrics data. 

To avoid these issues, follow these recommendations:

* Avoid querying endpoints frequently. Limit queries to a maximum of one every 30 seconds.
* Do not try to retrieve all metrics data via the `/federation` endpoint. Query it only when you want to retrieve a limited, aggregated data set. For example, retrieving fewer than 1,000 samples for each request helps minimize the risk of performance degradation.
* [What else?]
